cpp/src/copying/contiguous_split.cu

```diff
@@ -1158,116 +1286,700 @@ std::vector<packed_table> contiguous_split(cudf::table_view const& input,
+    // them into much smaller batches in order to drive up the number of blocks and overall
+    // occupancy.
+    rmm::device_uvector<thrust::pair<std::size_t, std::size_t>> batches(num_bufs, stream, mr);
+    auto d_dst_buf_info     = partition_buf_size_and_dst_buf_info->d_dst_buf_info;
+    auto h_buf_sizes        = partition_buf_size_and_dst_buf_info->h_buf_sizes;
+    auto desired_batch_size = contiguous_split_state::desired_batch_size;
+    thrust::transform(
+      rmm::exec_policy(stream, mr),
+      d_dst_buf_info,
+      d_dst_buf_info + num_bufs,
+      batches.begin(),
+      [desired_batch_size] __device__(
+        dst_buf_info const& buf) -> thrust::pair<std::size_t, std::size_t> {
+        // Total bytes for this incoming partition
+        std::size_t const bytes =
+          static_cast<std::size_t>(buf.num_elements) * static_cast<std::size_t>(buf.element_size);
+
+        // This clause handles nested data types (e.g. list or string) that store no data in the row
+        // columns, only in their children.
+        if (bytes == 0) { return {1, 0}; }
+
+        // The number of batches we want to subdivide this buffer into
+        std::size_t const num_batches = std::max(
+          std::size_t{1}, util::round_up_unsafe(bytes, desired_batch_size) / desired_batch_size);
+
+        // NOTE: leaving batch size as a separate parameter for future tuning
+        // possibilities, even though in the current implementation it will be a
+        // constant.
+        return {num_batches, desired_batch_size};
+      });
 
-  // build the output.
-  std::vector<packed_table> result;
-  result.reserve(num_partitions);
-  std::vector<column_view> cols;
-  cols.reserve(num_root_columns);
-  auto cur_dst_buf_info = h_dst_buf_info;
-  for (std::size_t idx = 0; idx < num_partitions; idx++) {
-    // traverse the buffers and build the columns.
-    cur_dst_buf_info = build_output_columns(
-      input.begin(), input.end(), cur_dst_buf_info, std::back_inserter(cols), h_dst_bufs[idx]);
-
-    // pack the columns
-    cudf::table_view t{cols};
-    result.push_back(packed_table{
-      t,
-      packed_columns{
-        std::make_unique<std::vector<uint8_t>>(cudf::pack_metadata(
-          t, reinterpret_cast<uint8_t const*>(out_buffers[idx].data()), out_buffers[idx].size())),
-        std::make_unique<rmm::device_buffer>(std::move(out_buffers[idx]))}});
-
-    cols.clear();
+    chunk_iter_state = make_chunk_iteration_state(
+      batches, num_bufs, d_dst_buf_info, h_buf_sizes, num_partitions, user_buffer_size, stream, mr);
   }
-  return result;
+
+  std::vector<packed_table> contiguous_split()
+  {
+    CUDF_EXPECTS(user_buffer_size == 0, "Cannot contiguous split with a user buffer");
+    if (is_empty || input.num_columns() == 0) { return make_packed_tables(); }
+
+    std::size_t num_batches_total;
+    std::tie(std::ignore, num_batches_total) =
+      chunk_iter_state->get_current_starting_index_and_buff_count();
```

Comment:
structured bindings?

Reply:
I changed the other  places for structured bindings, but this one I am trying to use `std::ignore`. I can change it as a structured binding with an unused variable, but this felt cleaner. Let me know if there's a better approach here.

Reply:
I'm partial to this approach myself:
```
auto const num_batches_total = std::get<1>(chunk_iter_state->get_current_starting_index_and_buff_count());
```
This allows the variable to marked `const` too perhaps.

Reply:
will do!